sklearn/pipeline.py

```diff
@@ -20,12 +21,73 @@
 from .externals import six
 from .utils import tosequence
 from .utils.metaestimators import if_delegate_has_method
-from .externals.six import iteritems
 
 __all__ = ['Pipeline', 'FeatureUnion']
 
 
-class Pipeline(BaseEstimator):
+class _BasePipeline(six.with_metaclass(ABCMeta, BaseEstimator)):
+    """Handles parameter management for classifiers composed of named steps.
+    """
+
+    @abstractmethod
+    def __init__(self):
+        pass
+
+    def _replace_step(self, steps_attr, name, new_val):
+        new_steps = getattr(self, steps_attr)[:]
```

Comment:
With `[:]` you definitely make a copy. If you plan on removing/adding entries from a list you are iterating over you need a copy but in this case we just replace parts so might not be needed.